rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,350 @@
+
+S3 event can be use to trigger additional processing such as validation, once an upload has completed.
+
+### Cloud to Cloud Upload Flow
+
+Cloud to cloud uploads will allow a user to provide a sharable link from their CSP to the file they wish to upload,
+which DP will used to upload the file to our S3 bucket. Now to walk though the cloud to cloud flow, see figure 1.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 2:** Architecture of a user uploading from their cloud to our cloud.
+
+#### 1. Get Share Link
+
+The user will generate a sharable link from their CSP. Both Google Drive and Dropbox support the generation of
+sharable links. Only sharable links should only point to a single file, not a folder or collection of files.
+
+#### 2. Share Link
+
+Using the browser App, the user will make a _POST {collection_id}/{dataset_id}/upload/link_ request to upload the file
+by providing the sharable link in the request body. The response will contain the asset uuid, that the user can use to
+check the status of the upload
+
+#### 3. Sanitize
+
+To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
+incoming shared link will be sanitized before queuing for download. The links generated by CSP are often not direct
+links to the files and often need to be parsed into the correct shape for downloading. The correct URL will be generated
+as part of sanitization. The backend that runs this should also be given minimal privileges to further protect from
+server side request forgery.
+
+##### How to download from a shared link
+
+- Dropbox <https://help.dropbox.com/files-folders/share/force-download>
+- Google drive <https://www.labnol.org/internet/direct-links-for-google-drive/28356/>
+
+#### 4. _Optional_ Queue
+
+This part is optional and depends on load we expect and server type we select. The queue stores the links that need to
+be processed. It buffers the down stream servers from download requests. This allows us to spin up servers as needed.
+It also gives us the ability to retry a download if one of servers fails to download or upload the file.
+
+##### Why might the download fail
+
+- The downloading server ran out of storage.
+- The downloading server crashed.
+- The CSP experience and outage
+- The user has reached their daily bandwidth allowance from their CSP. [more info](https://help.dropbox.com/files-folders/share/banned-links)
+
+#### 5. Start EC2 Instance to Download File
+
+Spin up EC2 instances with enough storage to download the whole file. TODO: See if spot instances could be used.
+
+#### 6. Download the File
+
+The most time will be spent here, downloading the file. Most CSP offer a way to validate the integrity of your download 
+by providing a hash of the data. We can compare this hash with the hash of downloaded file. If the hash does not match 
+we can reattempt the download.
```

Comment:
How many times would the service reattempt the download? How does the REST API surface the file corruption?